routing/pathfind.go

```diff
@@ -684,6 +680,16 @@ func findPath(g *graphParams, r *RestrictParams, source, target Vertex,
 		// examine all the incoming edges (channels) from this node to
 		// further our graph traversal.
 		pivot := Vertex(bestNode.PubKeyBytes)
+
+		// But before that, we'll check that we are not visiting the same
+		// node again which can happen because of the way our pathfinding
+		// algorithm works. Fortunately, only the shortest path to this
+		// node will be evaluated and all subsequent iterations over
+		// this node can safely be ignored.
+		if _, ok := visitedNodes[pivot]; ok {
```

Comment:
I am also thinking about a different fix, where the node is actually replaced on the heap instead of added again. There already exists a `heap.Fix` function to reposition an item. Maybe a direction is to extend `distanceHeap` with another field that is a map to lookup the heap index by pubkey.

Reply:
Time complexity of `heap.Fix` is `O(log n)` as is `heap.Push`.  So I don't think we save any on time when replacing the node on the heap (we do save space).  Plus we would also still have the space requirement of having a map of heap indices.  Also it still wouldn't address the problem of re-iterating over previously visited nodes in the `bestNode.ForEachChannel` loop, so we would still need the `visitedNodes` map.  Thoughts?

Reply:
My previous comment was a little wrong. `heap.Fix` fixes having duplicate items on the heap and probably has a similar space constraint as the `visitedNodes` map.  The only problem I see with it is that the nodes in the `ForEachChannel` loop will be visited again.  Sorry about the inaccuracy.

Reply:
I think the problem with visiting again is that it needs that db lookup? Other than that it would be pretty fast, perhaps neglible to visit again?

This can also be fixed in a different way by just not doing the db lookup when fetching the other end of the channel. Pass in a `Vertex` without the associated db fields to `processEdge`. Then create a different `ForEachChannel` that just operates on a `tx` and a `Vertex`.

What do you think about that direction? It may also be an additional speed up if nodes are visited multiple times before ending up in the `visitedNodes` map.

Reply:
Hey I need some clarification on this.  Are you suggesting adding a `ForEachChannel` to the `Vertex` which takes a `tx` + callback function which emulates the current `ForEachChannel` being used?  This requires a change to the heap to only store `Vertex` right?  This is good because then there's no db lookup required in the callback itself before `processEdge`.  Combining this with `heap.Fix`, we wouldn't really need a `visitedNodes` map since there would be no db call there.

Reply:
Yes that is the suggestion.

Reply:
The new commits implement this, I tested with the graph mentioned in a previous comment.  The `pivot` node isn't added to the heap again.